main.go

```diff
@@ -453,25 +497,156 @@ func (c *controller) waitForCacheSync(stop <-chan struct{}) error {
 	return nil
 }
 
-func (c *controller) worker() {
-	for c.queue.get() {
+func (c *controller) podWorker() {
+	fn := func() bool {
+		key, quit := c.podQ.Get()
+		if quit {
+			return false
+		}
+		defer c.podQ.Done(key)
+		if err := c.cleanUp(key.(*corev1.Pod)); err != nil {
+			level.Error(c.logger).Log("msg", "unable to clean up PVC", "err", err)
+			c.podQ.AddRateLimited(key)
+			return true
+		}
+		c.podQ.Forget(key)
+		return true
+	}
+	for fn() {
+	}
+}
+
+func (c *controller) stsWorker() {
+	for c.stsQ.get() {
 		c.sync()
 	}
 }
 
+func (c *controller) resolvePodOwnerRef(namespace string, refs []metav1.OwnerReference) (*appsv1.StatefulSet, error) {
+	for _, ref := range refs {
+		// If the owner reference points at the wrong kind of object, skip.
+		if ref.Kind != "StatefulSet" {
+			continue
+		}
+		// If the owner reference points at something that we don't have, then skip.
+		obj, ok, err := c.ssetInf.GetStore().GetByKey(fmt.Sprintf("%s/%s", namespace, ref.Name))
+		if !ok {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		sts := obj.(*appsv1.StatefulSet)
+		if sts.UID != ref.UID {
+			return nil, errors.Wrap(err, "owner reference UID does not match StatefulSet")
+		}
+		return sts, nil
+	}
+	return nil, nil
+}
+
+func (c *controller) cleanUp(pod *corev1.Pod) error {
+	c.cleanupAttempts.Inc()
+	sts, err := c.resolvePodOwnerRef(pod.Namespace, pod.GetOwnerReferences())
+	if err != nil {
+		c.cleanupErrors.WithLabelValues(fetchLabel).Inc()
+		return errors.Wrap(err, "could not get StatefulSet")
+	}
+	// This probably means that the Pod did not belong to a StatefulSet with
+	// our label selector, i.e. not a StatefulSet we are watching.
+	if sts == nil {
+		return nil
+	}
+	// Nothing to clean up.
+	if len(sts.Spec.VolumeClaimTemplates) == 0 {
+		return nil
+	}
+
+	helper := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cleanup-" + pod.Name,
+		},
+		Spec: corev1.PodSpec{
+			RestartPolicy: corev1.RestartPolicyNever,
+			Containers: []corev1.Container{
+				{
+					Name:            "cleanup",
+					Image:           c.options.cleanupImage,
+					Command:         []string{"rm", "-rf"},
```

Comment:
Just `rm -rf` doesn't deliver what we want to achieve here. We want to ensure data eventually gets shipped appropriately, so this should run a one-off `thanos replicate`, type thing.